fabric-object-builder-api-v1/src/main/java/net/fabricmc/fabric/api/event/fluid/FluidFlowEvents.java

```diff
@@ -0,0 +1,101 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package net.fabricmc.fabric.api.event.fluid;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import org.jetbrains.annotations.Nullable;
+
+import net.minecraft.block.Block;
+import net.minecraft.block.BlockState;
+import net.minecraft.util.Pair;
+import net.minecraft.util.math.BlockPos;
+import net.minecraft.util.math.Direction;
+import net.minecraft.world.World;
+
+public final class FluidFlowEvents {
+	private FluidFlowEvents() {
+	}
+
+	private static final Map<Block, Map<Block, List<Pair<Direction[], FluidFlowInteractionEvent>>>> EVENT_MAP = new HashMap<>();
+
+	/**
+	 * Registers a new event on a fluid flow. The same two blocks can register a different event for different directions, but the same directions will run the event that was registered first.
+	 *
+	 * @param flowingBlock          The fluid block that flowed.
+	 * @param interactionBlock      The block in one of the {@code interactionDirections}.
+	 * @param interactionDirections The direction to search for {@code interactionBlock}.
+	 * @param interactionEvent      The event to run when the conditions are met.
+	 */
+	public static void register(Block flowingBlock, Block interactionBlock, Direction[] interactionDirections, FluidFlowInteractionEvent interactionEvent) {
+		Map<Block, List<Pair<Direction[], FluidFlowInteractionEvent>>> flowBlockEvents = EVENT_MAP.getOrDefault(flowingBlock, new HashMap<>());
+		List<Pair<Direction[], FluidFlowInteractionEvent>> interactionEvents = flowBlockEvents.getOrDefault(interactionBlock, new ArrayList<>());
+		interactionEvents.add(new Pair<>(interactionDirections, interactionEvent));
+
+		if (!flowBlockEvents.containsKey(interactionBlock)) {
+			flowBlockEvents.put(interactionBlock, interactionEvents);
+		}
+
+		if (!EVENT_MAP.containsKey(flowingBlock)) {
+			EVENT_MAP.put(flowingBlock, flowBlockEvents);
+		}
+	}
+
+	/**
+	 * Gets the event from the following blocks and direction.
+	 *
+	 * @param flowingBlock         The fluid block that flowed.
+	 * @param interactionBlock     The block it interacts with.
+	 * @param interactionDirection The interaction direction
+	 * @return An event if the conditions are met, otherwise {@code null}
+	 */
+	public static @Nullable FluidFlowInteractionEvent getEvent(Block flowingBlock, Block interactionBlock, Direction interactionDirection) {
```

Comment:
This is a bit concerning, IMO this should allow `Event`-like chaining of callbacks.